library(tidyverse)
library(patchwork)
library(Seurat)
library(viridis)
library(EnhancedVolcano)
library(UpSetR)
library(scico)
library(pheatmap)
library(RColorBrewer)
library(magrittr)
library(ggbeeswarm)


setwd("E:/OneDrive - University of Glasgow/2 Projects/2 scRNAseq/S BM B scRNAseq/3 Code")

datadir = "Raw data/"
input.filepath = "Analysis results/2 Analysis/"

# Color scales ----

col.BM <- c("#82393f", "#D6616B", "#9C9EDE", "#6B6ECF", "#393B79",
            "#A55194", "#DE9ED6","#59753e", "#94ab57", "#b38839",
            "#E7BA52", "#9c446d", "#705828")

col.S <- c("#D6616B", "#393B79", "#7B4173", "#CE6DBD", "#DE9ED6", "#94ab57", "#E7BA52")

col.stages <- c('#ccebc5','#a8ddb5','#7bccc4', '#4eb3d3','#0868ac','#034e7b',
                "#83409B","#80146E", "#FFC08E", "#EC5D2F","#C12131")

col.early <- c('#ccebc5','#a8ddb5','#7bccc4', '#4eb3d3','#0868ac', "#FFC08E")

col.Retic <- c('#e65053','#2c7bb6')

col.organs <- c("#2D3139", "#4B8F8C", "#A64253")

# Loading File ----
# Requires Lists of DEG generated by "Seurat Analysis S" and "Seurat Analysis BM"

  BM.combined <- readRDS(paste0(input.filepath, "BM.combined.rds"))
  S.combined <- readRDS(paste0(input.filepath, "S.combined.rds"))
  Pb.combined.noMCA <- readRDS(paste0(input.filepath, "Pb.combined.noMCA.rds"))
  Pb.combined <- readRDS(paste0(input.filepath, "Pb.combined.rds"))
  
  BM.InfvsNaive <- readRDS(paste0("Analysis results/3 DEG/", "BM.InfvsNaive.rds"))
  S.InfvsNaive <- readRDS(paste0("Analysis results/3 DEG/", "S.InfvsNaive.rds"))
  Pb.RvsN <- readRDS(paste0("Analysis results/3 DEG/", "Pb.RvsN.rds"))
  
# Host analysis
# Host UMAP ----

DimPlot(BM.combined, group.by = "BMIdent", cols = col.BM) & NoLegend()
DimPlot(BM.combined, group.by = "seurat_clusters", label = T) & NoLegend()

DimPlot(S.combined, group.by = "SIdent", cols = col.S) & NoLegend()
DimPlot(S.combined, group.by = "seurat_clusters", label = T) & NoLegend()

  # Host Feature Plots Marker genes ----

DefaultAssay(BM.combined) <- "RNA"

BM.Marker <- c("Cd34", "Hbb-bt", "Gata1", "Fcnb", "Ly6g", 
               "Vpreb1","Ebf1", "Cd79a", "Cd74", "Siglech",
               "Ly6c1", "Cd68", "Cd3d", "Ncr1")

BM <- FeaturePlot(BM.combined, BM.Marker, max.cutoff = "q99", order = T, cols = c("light grey", scico(100, palette = 'lajolla', end = 0.90), pt.size = 0.1)) & 
  NoAxes() & NoLegend() & theme(panel.background = element_rect(colour = "black", size = 0, fill = NA)) 
BM <- BM + plot_layout(ncol = 5)
BM 

S.Marker <- c("Hbb-bt", "Ly6g", "Cd68", "Cd79a", "Cd74",
               "Cd3d", "Cd8a", "Cd4", "Ncr1")
S <- FeaturePlot(S.combined, S.Marker, max.cutoff = "q99", order = T, cols = c("light grey", scico(100, palette = 'lajolla', end = 0.90))) & 
  NoAxes() & NoLegend() & theme(panel.background = element_rect(colour = "black", size = 0, fill = NA)) 
S <- S + plot_layout(ncol = 5)
S


# Plot to get the legend
  FeaturePlot(BM.combined, "Hba-a2", max.cutoff = "q99", order = T, cols = c("light grey", scico(100, palette = 'lajolla', end = 0.90))) & 
  NoAxes() & theme(panel.background = element_rect(colour = "black", size = 1, fill = NA))

  # Host DEG general Analysis ----

  BM.Up <- lapply(BM.InfvsNaive, function(u)  u %>% 
                 dplyr::filter(FDR < 0.01) %>% dplyr::filter(logFC > 0) %>% 
                 dplyr::arrange(FDR))

  BM.Down <- lapply(BM.InfvsNaive, function(u)  u %>% 
                   dplyr::filter(FDR < 0.01) %>% dplyr::filter(logFC < 0) %>% 
                   dplyr::arrange(FDR))

  S.Up <- lapply(S.InfvsNaive, function(u)  u %>% 
                   dplyr::filter(FDR < 0.01) %>% dplyr::filter(logFC > 0) %>% 
                   dplyr::arrange(FDR))
  
  S.Down <- lapply(S.InfvsNaive, function(u)  u %>% 
                   dplyr::filter(FDR < 0.01) %>% dplyr::filter(logFC < 0) %>% 
                   dplyr::arrange(FDR))
  
  upset(fromList(map(BM.Up, "gene_id")), sets = names(BM.Up), nintersects = NA, 
        set_size.show = T, 
        sets.x.label = "# Differentially expressed Genes",
        mainbar.y.label = "# Genes per intersection",
        set_size.scale_max = 400)
  
  upset(fromList(map(BM.Down, "gene_id")), sets = names(BM.Down), nintersects = NA, 
        set_size.show = T, 
        sets.x.label = "# Differentially expressed Genes",
        mainbar.y.label = "# Genes per intersection",
        set_size.scale_max = 500)
  
  upset(fromList(map(S.Up, "gene_id")), nintersects = NA, 
        set_size.show = T, 
        sets.x.label = "# Differentially expressed Genes",
        mainbar.y.label = "# Genes per intersection",
        set_size.scale_max = 250)
  
  upset(fromList(map(S.Down, "gene_id")), nintersects = NA, 
        set_size.show = T, 
        sets.x.label = "# Differentially expressed Genes",
        mainbar.y.label = "# Genes per intersection",
        set_size.scale_max = 70)

  # Host DEG individual genes ----
  
  # Spleen
  S.Clus <- c("B cells", "CD4+ T cells", "CD8+ T cells", "NK cells", "Macrophages", "Neutrophils")
  
  S.InfectedvsNaive <- do.call("rbind",  S.InfvsNaive)
  S.InfectedvsNaive %<>% subset(cluster %in% S.Clus & logFC > 0 & FDR < 0.1) %>% group_by(cluster) %>% arrange(FDR, by_group = TRUE) 
  S.Top5 <- slice_head(S.InfectedvsNaive, n = 5)$gene_id %>% unique()

  S.combined$SIdent.Rep <- paste0(S.combined$SIdent, "_", S.combined$Biorep)
  Idents(S.combined) <- "SIdent.Rep"
  S.Cluster <- c("Macrophages_1", "Macrophages_2", "Macrophages_3", "Neutrophils_1", "Neutrophils_2", "Neutrophils_3",
                 "B cells_1", "B cells_2", "B cells_3", "CD4+ T cells_1", "CD4+ T cells_2", "CD4+ T cells_3",
                 "CD8+ T cells_1", "CD8+ T cells_2", "CD8+ T cells_3", "NK cells_1", "NK cells_2", "NK cells_3")
  
  S.avg <- AverageExpression(S.combined)$RNA
  S.avg.log <- log1p(S.avg)
  
  breaksList = c(seq(0, 3.6, by = 0.1))
  S.avg.log %>% subset(rownames(S.avg.log) %in% S.Top5) %>%
    subset(select = S.Cluster) %>%
    as.matrix() %>% pheatmap(cluster_cols = FALSE, 
                            color = c(colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(length(breaksList) - 1), "purple"), 
                            breaks = breaksList)
  
  VlnPlot(subset(S.combined, SIdent %in% S.Clus), features = S.Top5, group.by = "SIdent", split.by = "Biorep", pt.size = 0, 
          cols = c("dark red", "dark red", "grey")) & theme(axis.text = element_text(size = 8), 
                                                            title = element_text(size = 8))
  
  DotPlot(S.combined, features = S.Top5, idents = S.Cluster, group.by = "SIdent.Rep",
          cols = "RdYlBu") & theme(axis.text = element_text(size = 8), 
                                   title = element_text(size = 8))
  
  # Bone Marrow
  BM.Clus <- c("B cells", "T cells", "DCs", "NK cells", "Macrophages", "Neutrophils")
  
  BM.InfectedvsNaive <- do.call("rbind",  BM.InfvsNaive)
  BM.InfectedvsNaive %<>% subset(cluster %in% BM.Clus & logFC > 0 & FDR < 0.1) %>% group_by(cluster) %>% arrange(FDR, by_group = TRUE)
  BM.Top5 <- slice_head(BM.InfectedvsNaive, n = 5)$gene_id %>% unique()

  BM.combined$BMIdent.Rep <- paste0(BM.combined$BMIdent, "_", BM.combined$Biorep)
  Idents(BM.combined) <- "BMIdent.Rep"
  BM.Cluster <- c("Macrophages_1", "Macrophages_2", "Macrophages_3",
                  "Neutrophils_1", "Neutrophils_2", "Neutrophils_3",
                  "B cells_1", "B cells_2", "B cells_3", 
                  "T cells_1", "T cells_2", "T cells_3", "DCs_1", "DCs_2",
                  "DCs_3", "NK cells_1","NK cells_2", "NK cells_3")

  BM.avg <- AverageExpression(BM.combined)$RNA
  BM.avg.log <- log1p(BM.avg)
  
  BM.avg.log %>% subset(rownames(BM.avg.log) %in% BM.Top5) %>%
    subset(select = BM.Cluster) %>%
    as.matrix() %>% pheatmap(cluster_cols = FALSE, scale = "none")
  
  breaksList = c(seq(0, 3.6, by = 0.1))
  BM.avg.log %>% subset(rownames(BM.avg.log) %in% BM.Top5) %>%
    subset(select = BM.Cluster) %>%
    as.matrix() %>% pheatmap(cluster_cols = FALSE, 
                             color = c(colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(length(breaksList) - 1), "purple"), 
                             breaks = breaksList)
  
  breaksList = c(seq(0, 4, by = 0.1), seq(7,9, by = 0.1))
  BM.avg.log %>% subset(rownames(BM.avg.log) %in% BM.Top5) %>%
    subset(select = BM.Cluster) %>%
    as.matrix() %>% pheatmap(cluster_cols = FALSE, 
                             color = c(colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(45),
                                       "white",
                                       colorRampPalette(brewer.pal(n = 7, name = "RdPu"))(20)), 
                             breaks = breaksList)


  
  VlnPlot(subset(BM.combined, BMIdent %in% BM.Clus), features = BM.Top5, group.by = "BMIdent", split.by = "Biorep", pt.size = 0, 
          cols = c("dark red", "dark red", "grey")) & theme(axis.text = element_text(size = 8), 
                                                            title = element_text(size = 8))
  
  DotPlot(BM.combined, features = BM.Top5, idents = BM.Cluster, group.by = "BMIdent.Rep",
          cols = "RdYlBu") & theme(axis.text = element_text(size = 8), 
                                                            title = element_text(size = 8))
 
# Parasite analysis ---- 
  # Parasite UMAP ----

  p1 <- DimPlot(subset(Pb.combined, Organ == "MCA", invert = T), group.by = "seurat_clusters", label = T) & NoLegend()
  p2 <- DimPlot(subset(Pb.combined, Organ == "MCA"), group.by = "absclust3", label = T) & NoLegend()
  
  p1 + p2
  
  heatmap(table(Pb.combined$seurat_clusters, Pb.combined$absclust3), Colv = NA, Rowv = NA) 
  
  VlnPlot(subset(Pb.combined, Organ == "MCA", invert = T), "PbUMI1", group.by = "seurat_clusters", pt.size = 0) + scale_y_log10()
  
  DimPlot(Pb.combined.noMCA, group.by = "Stages", cols = col.stages, pt.size = 2) & NoLegend()
  
  DimPlot(Pb.combined.noMCA, split.by = "Organ", group.by = "Stages", cols = col.stages, pt.size = 2) & NoLegend()
  
  DimPlot(subset(Pb.combined.noMCA, Organ == "BM"), group.by = "Stages", cols = col.stages, pt.size = 2) & NoLegend()
  DimPlot(subset(Pb.combined.noMCA, Organ == "S"), group.by = "Stages", cols = col.stages, pt.size = 2) & NoLegend()
  DimPlot(subset(Pb.combined.noMCA, Organ == "B"), group.by = "Stages", cols = col.stages, pt.size = 2) & NoLegend()
  
  
  clipr::write_clip(table(Pb.combined$Stages, Pb.combined$orig.ident))
  
  # Organ DEG heatmaps ----
  
  Pb.combined.noMCA$Stages.Rep <- paste0(Pb.combined.noMCA$Stages, "_", Pb.combined.noMCA$orig.ident)
  Idents(Pb.combined.noMCA) <- "Stages.Rep"
  
  Pb.avg <- AverageExpression(Pb.combined.noMCA)$RNA
  Pb.avg.log <- log1p(Pb.avg)

  Pb.Clus <-  c("Late Trophs")
  
  tt.organs <- readRDS(paste0("Analysis results/3 DEG/", "Pb.organs.tt.rds"))
  Pb.Marker <- tt.organs %>% dplyr::filter(FDR < 0.05) %>% dplyr::arrange(FDR)
  Pb.Marker %<>% subset(cluster %in% Pb.Clus) %>% group_by(contrast) %>% arrange(FDR, .by_group = TRUE)
  Top <- slice_head(Pb.Marker, n = 25)$gene_id %>% unique()

  breaksList = c(seq(0, 3, by = 0.1))
  Pb.avg.log %>% subset(rownames(Pb.avg.log) %in% Top) %>%
    subset(select = c("Late Trophs_S1", "Late Trophs_S2", "Late Trophs_BM1", "Late Trophs_BM2", "Late Trophs_B1", "Late Trophs_B2")) %>%
    as.matrix() %>% pheatmap(cluster_cols = FALSE,
                             color = c(colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(length(breaksList) - 1), "purple"),
                             breaks = breaksList)
  

  # CD71 and retic analysis ----
  
  clipr::write_clip(table(Pb.combined.noMCA$CD71_Retic, Pb.combined.noMCA$orig.ident))
  clipr::write_clip(Pb.combined.noMCA[[]] %>% unite(col = sample, orig.ident, Stages) %>% 
      group_by(sample) %>% select(CD71pos) %>% count(CD71pos) %>% spread(CD71pos, n) %>%
      summarise(CD71 = 100*pos/(pos+neg))%>% 
      separate(sample, into = c("orig.ident", "Stages"), sep = "_") %>% spread(Stages, CD71))
      
  clipr::write_clip(Pb.combined.noMCA[[]] %>% unite(col = sample, orig.ident, Stages) %>% 
      group_by(sample) %>% select(CD71pos) %>% count(CD71pos) %>% 
        separate(sample, into = c("orig.ident", "Stages"), sep = "_") %>% spread(CD71pos, n))
      
  FeaturePlot(Pb.combined.noMCA, "CD71", pt.size = 2, max.cutoff = "q99", order = F, 
              cols = c(scico(100, palette = 'lajolla'))) & NoLegend() + 
              theme(plot.title = element_blank())
  
  DimPlot(Pb.combined.noMCA, group.by = "Retic", cols = col.Retic, pt.size = 2) & NoLegend() + 
    theme(plot.title = element_blank())
  DimPlot(Pb.combined.noMCA, group.by = "Stages", split.by = "Retic", cols = col.stages, pt.size = 2) & NoLegend() + 
    theme(plot.title = element_blank())
  
  
  PlotMmUMI <- Pb.combined.noMCA[[c("Stages", "MmUMI1", "CD71pos")]]
  CD71 <- Pb.combined.noMCA@assays$ADT@data[row.names(Pb.combined.noMCA@assays$ADT@data) %in% "CD71",]
  PlotMmUMI$CD71 <- CD71
  
  ggplot(PlotMmUMI, aes(x = CD71, y = MmUMI1)) + 
    scale_y_log10() +
    geom_jitter() +
    stat_density_2d(aes(fill = ..level..), geom = "polygon", colour = "black") +
    scale_fill_continuous(type = "viridis") + 
    geom_hline(yintercept = 100) + 
    geom_vline(xintercept = 1.5) + 
    theme_classic()

  Pb.subset <- subset(Pb.combined.noMCA, downsample = 500)
  CD71 <- Pb.subset@assays$ADT@data[row.names(Pb.subset@assays$ADT@data) %in% "CD71",]
  PlotMmUMI2 <- Pb.subset[[c("Stages", "MmUMI1", "CD71pos", "Retic")]]
  PlotMmUMI2$CD71 <- CD71

  ggplot(PlotMmUMI2, aes(x = Stages, y = MmUMI1, fill = Stages)) + 
    geom_violin(trim = T, scale = "width") + scale_y_log10() + 
    scale_fill_manual(values = col.stages) + 
    geom_quasirandom(size = 1, aes(color = CD71pos)) + 
    scale_colour_manual(values = c("black", "violetred1")) + 
    geom_hline(yintercept = 100) + 
    theme_classic()
  
  ggplot(PlotMmUMI2, aes(x = Stages, y = MmUMI1, fill = Retic)) + 
    geom_violin(trim = T, scale = "width") + scale_y_log10() + 
    scale_fill_manual(values = col.Retic) + 
    geom_jitter(height = 0, width = 0.2, (aes(colour = Retic))) +
    geom_hline(yintercept = 100) + 
    theme_classic()
  
  Pb.retic <- subset(Pb.combined.noMCA,subset = "Retic" == Retic)
  
  PlotMmUMI.R <- Pb.retic[[c("Stages", "MmUMI1", "PbUMI")]]
  
  ggplot(PlotMmUMI.R, aes(x = Stages, y = MmUMI1, fill = Stages)) + 
    geom_violin(trim = T, scale = "width") + scale_y_log10() + 
    geom_quasirandom(size = 0.5) +
    scale_fill_manual(values = col.stages) +
    theme_classic()
  
  clipr::write_clip(
    Pb.combined.noMCA[[]] %>%  rownames_to_column() %>%
      group_by(Stages) %>% select(rowname, Stages, MmUMI1) %>% 
      spread(Stages, MmUMI1) %>% select(-c(rowname))
  )
  
  clipr::write_clip(
    subset(Pb.combined.noMCA,subset = "Retic" == Retic)[[]] %>%  rownames_to_column() %>%
    group_by(Stages) %>% select(rowname, Stages, MmUMI1) %>% 
      spread(Stages, MmUMI1) %>% select(-c(rowname))
    )
  
  clipr::write_clip(
    subset(Pb.combined.noMCA,subset = MmUMI < 100)[[]] %>%  rownames_to_column() %>%
      group_by(Stages) %>% select(rowname, Stages, MmUMI1) %>% 
      spread(Stages, MmUMI1) %>% select(-c(rowname))
  )
  
  clipr::write_clip(
  subset(Pb.combined.noMCA,subset = "Retic" == Retic)[[]] %>%  rownames_to_column() %>%
    unite("orig_Stage", orig.ident, Stages) %>% group_by(orig_Stage) %>% summarize(mean(MmUMI)) %>%
    separate(orig_Stage, c("orig.ident", "Stages"), sep = "_") %>% spread(Stages, 'mean(MmUMI)')
  )
  
  
  # DEG Retic vs Normo ----  
    
    Pb.Early <- subset(Pb.combined.noMCA, Stages == "Schizonts" | Stages == "Late Trophs" | 
                         Stages == "Mid Trophs" | Stages == "Males" | Stages == "Females", invert = TRUE)
    
    DefaultAssay(Pb.Early) <- "RNA"
    Idents(Pb.Early) <- "Stages"
    
    # Retic heatmap ----
  
    # Volcano Plots ----
  
  Pb.RvsN <- readRDS(paste0("Analysis results/3 DEG/", "Pb.RvsN.rds"))

  GOI <- c("PBANKA-1437500.1", "PBANKA-1113000.1", "PBANKA-1210800.1", "PBANKA-0513600.1",
          "PBANKA-0306000.1", "PBANKA-1204400.1", "PBANKA-1128300.1", "PBANKA-0817900.1", "PBANKA-1360100.1")
    
  RvsN.Volc <- lapply(names(Pb.RvsN), function(k){
    EnhancedVolcano(Pb.RvsN[[k]], 
                    lab = Pb.RvsN[[k]]$gene_id, 
                    x = "logFC", y = "FDR",
                    selectLab = GOI,
                    pCutoff = 0.05, FCcutoff = 0, 
                    xlim = c(-2.2,2.2), ylim = c(0, 6), 
                    xlab = bquote(~Log[2]~ "fold change"),
                    ylab = bquote(~-Log[10]~ "FDR"), 
                    title = names(Pb.RvsN[k]),
                    subtitle = "",
                    drawConnectors = TRUE,
                    boxedLabels = TRUE,
                    legendPosition = "top")
  })

  
  p1 <- RvsN.Volc[[1]]
  p2 <- RvsN.Volc[[2]]
  p3 <- RvsN.Volc[[3]]
  p4 <- RvsN.Volc[[4]]
  p5 <- RvsN.Volc[[5]]
  p6 <- RvsN.Volc[[6]]
  
  p6 + p5 + p3
 
  # Purine analysis ----
 
  purines <- c("PBANKA-1113000.1", "PBANKA-1210800.1", "PBANKA-0513600.1")
  VlnPlot(subset(Pb.Early, Stages == "Early Gams", invert = T), purines, 
          group.by = "Stages", split.by = "Retic", cols = col.Retic, pt.size = 0)
  
  # AP2G analysis ----
  
  AP2G <- "PBANKA-1437500.1"
  
  FeaturePlot(Pb.combined.noMCA, AP2G, max.cutoff = "q90", order = T, pt.size = 2,
              cols = c("light grey", scico(99, palette = 'lajolla', end = 0.9))) & 
              NoLegend() & theme(plot.title = element_blank())
  
  # Percentage AP2G positive per host cell and organ
  
  pct <- readRDS(paste0(input.filepath, "Pb.pct.rds"))
  
  AP2G.pct <- filter(pct, GeneID == "PBANKA-1437500.1") %>% group_by(Stage) %>% gather(RBC, pct.exp, Normo, Retic)
  head(AP2G.pct)
  AP2G.pct$Stage <- factor(AP2G.pct$Stage, levels = c("Early Rings", "Mid Rings", "Late Rings", "Early Trophs 1", 
                                                      "Early Trophs 2", "Early Gams"))
  
  AP2G.pct$Organ <- factor(AP2G.pct$Organ, levels = c("S", "BM", "B"))
  
  ggplot(AP2G.pct, aes(y = pct.exp, x = Stage, fill = RBC)) +
    geom_boxplot() + 
    scale_fill_manual(values = c('#e65053','#2c7bb6')) +
    geom_jitter(shape = 16, position = position_jitterdodge(0)) +
    theme_classic()
  
  AP2G.R <- filter(AP2G.pct, RBC == "Retic")
  AP2G.N <- filter(AP2G.pct, RBC == "Normo")
  
  p1 <-  ggplot(AP2G.R, aes(y = pct.exp, x = Stage, fill = Organ)) +
    geom_boxplot() + 
    scale_fill_manual(values = col.organs) +
    ylim(0, 100) +
    geom_jitter(shape = 16, position = position_jitterdodge(0)) +
    theme_classic()
  
  p2 <- ggplot(AP2G.N, aes(y = pct.exp, x = Stage, fill = Organ)) +
    geom_boxplot() + 
    scale_fill_manual(values = col.organs) + 
    ylim(0, 100) +
    geom_jitter(shape = 16, position = position_jitterdodge(0)) +
    theme_classic()
  
  p1 + p2
  
  # AP2G.pct$Org_RBC <- paste0(AP2G.pct$RBC, "_", AP2G.pct$Organ)
  # 
  # ggplot(AP2G.pct, aes(y = pct.exp, x = Stage, fill = Org_RBC)) +
  #   geom_boxplot() + 
  #   geom_jitter(shape = 16, position = position_jitterdodge(0)) +
  #   theme_classic()
  # 
  # pct.all <- readRDS(paste0(input.filepath, "Pb.pct.all.rds"))
  # 
  # AP2G.pct.all <- filter(pct.all, GeneID == "PBANKA-1437500.1") %>% group_by(Stage) # %>% gather(RBC, pct.exp, Normo, Retic)
  # head(AP2G.pct.all)
  # AP2G.pct.all$Stage <- factor(AP2G.pct.all$Stage, levels = c("Early Rings", "Mid Rings", "Late Rings", "Early Trophs 1", 
  #                                                     "Early Trophs 2", "Early Gams"))
  # 
  # ggplot(AP2G.pct.all, aes(y = pct.exp, x = Stage, fill = Organ)) +
  #   geom_boxplot(outlier.shape = NA) + 
  #   geom_point(aes(fill = Organ), position = position_jitterdodge(0)) +
  #   theme_classic()
  # 
  
  
    
  
  